File: 二叉树/_平衡二叉树.go
package 二叉树

import "math"

func isBalanced(node *TreeNode) bool {
	if node == nil {
		return true
	}
	if !isBalanced(node.Left) || !isBalanced(node.Right) {
		return false
	}
	//后序遍历的单层逻辑，从下往上判断
	if math.Abs(float64(maxDeath(node.Left)-maxDeath(node.Right))) > 1 {
		return false
	}
	return true
}

func maxDeath(node *TreeNode) int {
	if node == nil {
		return 0
	}
	left := maxDeath(node.Left)
	right := maxDeath(node.Right)
	return max(left, right) + 1
}

File: 二叉树/平衡二叉树.java
package 二叉树;

public class 平衡二叉树 {
    public boolean isBalanced(TreeNode root) {
        if (root == null) {
            return true;
        }
        int left = maxdepth(root.left);
        int right = maxdepth(root.right);
        if (Math.abs(left - right) > 1) return false;
        return isBalanced(root.left) && isBalanced(root.right);
    }

    public int maxdepth(TreeNode root) {
        if (root == null) return 0;
        return Math.max(maxdepth(root.left), maxdepth(root.right)) + 1;
    }
}
